File: backend/routers/admin_users.py
"""
admin_users router — /api/admin-users CRUD + reset-password.
Super-admin only.
"""
from typing import Optional, List, Dict, Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, EmailStr, Field

from deps import (
    db,
    admin_user_shape,
    gen_id,
    hash_password,
    now_iso,
    require_super_admin,
    sanitize_input,
    verify_admin,
)

router = APIRouter()


# ---- Models ---------------------------------------------------------------
class AdminUserInviteIn(BaseModel):
    email: EmailStr
    name: str = Field(..., min_length=1, max_length=120)
    temp_password: str = Field(..., min_length=8, max_length=200)


class CustomFieldIn(BaseModel):
    label: str = Field("", max_length=60)
    value: str = Field("", max_length=300)


class AdminUserUpdateIn(BaseModel):
    name: Optional[str] = None
    designation: Optional[str] = None
    photo_url: Optional[str] = None
    initials: Optional[str] = None
    booking_url: Optional[str] = None
    show_as_host: Optional[bool] = None
    custom_fields: Optional[List[CustomFieldIn]] = None


class AdminUserResetIn(BaseModel):
    new_temp_password: str = Field(..., min_length=8, max_length=200)


# ---- Routes ---------------------------------------------------------------
@router.get("/api/admin-users")
async def admin_users_list(claims: Dict[str, Any] = Depends(verify_admin)):
    me = await require_super_admin(claims)
    rows = await db.admin_users.find({"email": {"$exists": True}}, {"_id": 0}).sort("created_at", 1).to_list(length=200)
    return {
        "success": True,
        "users": [admin_user_shape(r, include_temp_password=True) for r in rows],
        "me": admin_user_shape(me, include_temp_password=True),
    }


@router.post("/api/admin-users")
async def admin_user_invite(payload: AdminUserInviteIn, claims: Dict[str, Any] = Depends(verify_admin)):
    await require_super_admin(claims)
    email = payload.email.lower().strip()
    if not email.endswith("@orgainse.com"):
        raise HTTPException(400, "Only @orgainse.com emails may be invited")
    if await db.admin_users.find_one({"email": email}):
        raise HTTPException(400, "User with this email already exists")
    name = sanitize_input({"n": payload.name})["n"][:120]
    doc = {
        "id": gen_id(),
        "email": email,
        "name": name,
        "password_hash": hash_password(payload.temp_password),
        "temp_password_plain": payload.temp_password,
        "must_change_password": True,
        "role": "admin",
        "is_super_admin": False,
        "created_at": now_iso(),
        "last_login_at": None,
    }
    await db.admin_users.insert_one(doc)
    return {"success": True, "user": admin_user_shape(doc, include_temp_password=True)}


@router.put("/api/admin-users")
async def admin_user_update(
    payload: AdminUserUpdateIn,
    id: str = Query(...),
    claims: Dict[str, Any] = Depends(verify_admin),
):
    await require_super_admin(claims)
    user = await db.admin_users.find_one({"id": id})
    if not user:
        raise HTTPException(404, "User not found")
    updates: Dict[str, Any] = {"updated_at": now_iso()}
    if payload.name is not None:
        updates["name"] = sanitize_input({"n": payload.name})["n"][:120]
    if payload.designation is not None:
        updates["designation"] = sanitize_input({"d": payload.designation})["d"][:160]
    if payload.photo_url is not None:
        updates["photo_url"] = sanitize_input({"p": payload.photo_url})["p"][:1000]
    if payload.initials is not None:
        updates["initials"] = sanitize_input({"i": payload.initials})["i"][:4].upper()
    if payload.booking_url is not None:
        updates["booking_url"] = sanitize_input({"b": payload.booking_url})["b"][:500]
    if payload.show_as_host is not None:
        updates["show_as_host"] = bool(payload.show_as_host)
    if payload.custom_fields is not None:
        cleaned: List[Dict[str, str]] = []
        for cf in payload.custom_fields:
            label = sanitize_input({"l": cf.label})["l"][:60]
            value = sanitize_input({"v": cf.value})["v"][:300]
            if label:
                cleaned.append({"label": label, "value": value})
        updates["custom_fields"] = cleaned
    await db.admin_users.update_one({"id": id}, {"$set": updates})
    updated = await db.admin_users.find_one({"id": id}, {"_id": 0})
    return {"success": True, "user": admin_user_shape(updated, include_temp_password=True)}


@router.post("/api/admin-users/reset-password")
async def admin_user_reset_password(
    payload: AdminUserResetIn,
    id: str = Query(...),
    claims: Dict[str, Any] = Depends(verify_admin),
):
    await require_super_admin(claims)
    user = await db.admin_users.find_one({"id": id})
    if not user:
        raise HTTPException(404, "User not found")
    await db.admin_users.update_one(
        {"id": id},
        {"$set": {
            "password_hash": hash_password(payload.new_temp_password),
            "temp_password_plain": payload.new_temp_password,
            "must_change_password": True,
            "updated_at": now_iso(),
        }},
    )
    updated = await db.admin_users.find_one({"id": id}, {"_id": 0})
    return {"success": True, "user": admin_user_shape(updated, include_temp_password=True)}


@router.delete("/api/admin-users")
async def admin_user_delete(id: str = Query(...), claims: Dict[str, Any] = Depends(verify_admin)):
    me = await require_super_admin(claims)
    user = await db.admin_users.find_one({"id": id})
    if not user:
        raise HTTPException(404, "User not found")
    if user.get("id") == me.get("id") or user.get("is_super_admin"):
        raise HTTPException(400, "Cannot delete a super-admin or yourself")
    await db.admin_users.delete_one({"id": id})
    return {"success": True, "deleted": id}

File: api/_handlers/admin-users.js
/**
 * /api/admin-users        — GET list, POST invite, PUT update, DELETE
 * /api/admin-users/reset-password?id=...  — POST (handled by the explicit
 *                                            file at /api/admin-users/reset-password.js)
 *
 * Super-admin only. Mirrors FastAPI admin-users CRUD.
 */
import { securityHeaders, validateRequestSize, sanitizeInput } from '../_middleware/security.js';
import { requireAdmin } from '../_middleware/verify-admin.js';
import {
  getDb,
  hashPassword,
  requireSuperAdmin,
  adminUserShape,
  cryptoRandomId,
} from '../_auth-utils.js';

export default async function handler(req, res) {
  securityHeaders(req, res);
  if (req.method === 'OPTIONS') return res.status(200).end();
  if (!validateRequestSize(req, res, 64 * 1024)) return;

  const claims = requireAdmin(req, res);
  if (!claims) return;

  try {
    const db = await getDb();
    const me = await requireSuperAdmin(db, claims, res);
    if (!me) return;

    const { id } = req.query || {};

    // ---- GET: list users ----
    if (req.method === 'GET') {
      const rows = await db.collection('admin_users')
        .find({ email: { $exists: true } })
        .sort({ created_at: 1 })
        .toArray();
      return res.status(200).json({
        success: true,
        users: rows.map((r) => adminUserShape(r, { includeTempPassword: true })),
        me: adminUserShape(me, { includeTempPassword: true }),
      });
    }

    // ---- POST: invite new admin ----
    if (req.method === 'POST') {
      const body = sanitizeInput(req.body || {});
      const email = (body.email || '').toString().trim().toLowerCase();
      const name = (body.name || '').toString().trim();
      const tempPassword = (body.temp_password || '').toString();
      if (!email || !email.endsWith('@orgainse.com')) {
        return res.status(400).json({ error: 'Only @orgainse.com emails may be invited' });
      }
      if (!name) return res.status(400).json({ error: 'Name is required' });
      if (tempPassword.length < 8) {
        return res.status(400).json({ error: 'Temp password must be at least 8 chars' });
      }
      const existing = await db.collection('admin_users').findOne({ email });
      if (existing) return res.status(400).json({ error: 'User with this email already exists' });
      const doc = {
        id: cryptoRandomId(),
        email,
        name: name.slice(0, 120),
        password_hash: hashPassword(tempPassword),
        temp_password_plain: tempPassword,
        must_change_password: true,
        role: 'admin',
        is_super_admin: false,
        created_at: new Date().toISOString(),
        last_login_at: null,
      };
      await db.collection('admin_users').insertOne(doc);
      return res.status(200).json({ success: true, user: adminUserShape(doc, { includeTempPassword: true }) });
    }

    // ---- PUT: update user (name + host profile) ----
    if (req.method === 'PUT') {
      if (!id) return res.status(400).json({ error: 'id required' });
      const body = sanitizeInput(req.body || {});
      const updates = { updated_at: new Date().toISOString() };
      if (typeof body.name === 'string') updates.name = body.name.slice(0, 120);
      if (typeof body.designation === 'string') updates.designation = body.designation.slice(0, 160);
      if (typeof body.photo_url === 'string') updates.photo_url = body.photo_url.slice(0, 1000);
      if (typeof body.initials === 'string') updates.initials = body.initials.slice(0, 4).toUpperCase();
      if (typeof body.booking_url === 'string') updates.booking_url = body.booking_url.slice(0, 500);
      if (typeof body.show_as_host === 'boolean') updates.show_as_host = body.show_as_host;
      if (Array.isArray(body.custom_fields)) {
        updates.custom_fields = body.custom_fields
          .filter((cf) => cf && typeof cf.label === 'string' && cf.label.trim())
          .map((cf) => ({
            label: String(cf.label).slice(0, 60),
            value: String(cf.value || '').slice(0, 300),
          }));
      }
      const user = await db.collection('admin_users').findOne({ id });
      if (!user) return res.status(404).json({ error: 'User not found' });
      await db.collection('admin_users').updateOne({ id }, { $set: updates });
      const updated = await db.collection('admin_users').findOne({ id });
      return res.status(200).json({ success: true, user: adminUserShape(updated, { includeTempPassword: true }) });
    }

    // ---- DELETE: remove admin ----
    if (req.method === 'DELETE') {
      if (!id) return res.status(400).json({ error: 'id required' });
      const user = await db.collection('admin_users').findOne({ id });
      if (!user) return res.status(404).json({ error: 'User not found' });
      if (user.id === me.id || user.is_super_admin) {
        return res.status(400).json({ error: 'Cannot delete a super-admin or yourself' });
      }
      await db.collection('admin_users').deleteOne({ id });
      return res.status(200).json({ success: true, deleted: id });
    }

    return res.status(405).json({ error: 'Method not allowed' });
  } catch (err) {
    console.error('admin-users error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}
